### PR TITLE
Omit optional transformations in an opt marked as MustBeDone

### PR DESCRIPTION
**Pull request by liqunl**
Opts marked as MustBeDone will always run. However, the optional
transformations should be omitted if the opt's opt index is not in the
range specified by firstOptIndex and lastOptIndex.

Signed-off-by: Liqun Liu <liqunl@ca.ibm.com>

### CODE REVIEW

**Comment by liqunl** [comment]
This PR is part of the fix to address the issue where optimization searches is broken due to OSR infrastructure. This problem has increased the difficulty in defect investigation such as https://github.com/eclipse/openj9/issues/7684

@andrewcraik May I have your review?

File: compiler/ras/Debug.cpp [hunk 1]
```diff
@@ -492,6 +492,16 @@ TR_Debug::getDiagnosticFormat(const char *format, char *buffer, int32_t length)
...
    {
+   int32_t optIndex = _comp->getOptIndex();
+   int32_t firstOptIndex = _comp->getOptions()->getFirstOptIndex();
+   int32_t lastOptIndex = _comp->getOptions()->getLastOptIndex();
+   // An opt that is marked as MustBeDone will always run.
+   // However, we need to omit optional transformations when the opt is supposed
+   // to be prohibited by firstOptIndex and lastOptIndex
+   //
+   if (canOmitTransformation && (optIndex < firstOptIndex || optIndex > lastOptIndex))
+      return false;
```
Comment (fjeremic) [inline compiler/ras/Debug.cpp]: Does returning here interfere with the `TR_CountOptTransformations` code right below? It seems we may not execute `_comp->incOptSubIndex()`, etc. if we return early. Just making sure this is what we want.

Comment (liqunl) [reply compiler/ras/Debug.cpp]: Yes, this is what we want. If all optional transformations are prohibited by firstOptIndex and lastOptIndex, we don't need to increment sub index and transformation index. Sub index and transformation index are other ways to search bad transformations, but they're only used when the opt index has passed the criteria.

**Comment by andrewcraik** [comment]
@genie-omr build all

**Comment by fjeremic** [comment]
@genie-omr build zos

**Comment by andrewcraik** [comment]
@genie-omr build zos

**Comment by fjeremic** [comment]
Who's PR trigger will be the first to succeed 😂 

Filip 1 - 1 Andrew

@genie-omr build zos

**Comment by andrewcraik** [comment]
GIven the committers discussion around the issues on the zos machines (failure is infrastructure) and the fact this is not a z codegen or fvtest change I am going to merge this since the zos infrastructure is still proving quite flaky.

**Pull request merged**